### ISSUE TITLE
[FEATURE] Add linter to CI/CD pipeline 

### ISSUE THREAD

**Issue by garawalid** [scikit-multiflow/scikit-multiflow#219]
**Is your feature request related to a problem? Please describe.**
Actually, for each PR, the reviewer checks manually if the style PEP8 is correct. 

**Describe the solution you'd like**
It's possible to use `flake8` to check the PEP8 style. We integrate it into the CI/CD pipeline.

We still need to agree on some PEP8 rules:
- Do we keep max-line-length 120 (the same we do now) or we use 79 (as suggested by [PEP8](https://www.python.org/dev/peps/pep-0008/#id19))? 
- We need to ignore [E741](https://www.flake8rules.com/rules/E741.html) since we use a lot `x` and `y`.

**Comment by garawalid** [comment]
Related to #213

**Comment by garawalid** [comment]
I made a quick test and I found only these two style exceptions. Any thoughts?
@jacobmontiel @smastelini

**Comment by jacobmontiel** [comment]
Thanks for creating the issue @garawalid 

My two cents:

* Max line length
   This is a hot topic, and most of the time goes down to team preferences [(E501)](https://www.flake8rules.com/rules/E501.html). Personally, I find 79 characters rather restrictive. We could consider a middle ground at 99. Additionally, we need to consider the number of lines to change as we consider smaller line lengths.

* CapitalizedWords or single upper case letter
   I understand what you mean. I am not sure if it fits into E741 though. In this case, we are following a scikit-learn convention that is familiar to a large user base so I vote to ignore this rule.

**Comment by garawalid** [comment]
@jacobmontiel
I find 99 characters good as well, especially when I use a wide screen. 

I'll create an initial PR ASAP.

**Comment by smastelini** [comment]
Hi @garawalid and @jacobmontiel. I use a flake8-based plugin in my text editor, and I always received many warnings regarding the line lengths.  I also agree that 99 characters are a good compromise for our case.

@garawalid, this upcoming PR is going to be an excellent addition for the CI/CD pipeline! Thanks for that :)

**Comment by garawalid** [comment]
Using a 99-character rule will lead to many changes. I made a PR (#221) with an initial linting.  
I've been thinking of using 120 characters. What do you think?

**Comment by jacobmontiel** [comment]
> Using a 99-character rule will lead to many changes. I made a PR (#221) with an initial linting.
> I've been thinking of using 120 characters. What do you think?

I am ok with 120 characters, but I am biased. Could you provide an estimate of modules/files that would need to be changed if we use 99 characters? We could break the effort into multiple issues. We could start with a 120 character threshold for existing code and 99 characters for new code. Comments?

**Comment by garawalid** [comment]
Here are some statistics on the number of lines that exceed 99 character per module
```
# tests
      3 neural_networks
     16 data
      1 transform
     11 meta
      2 utils
      5 metrics
     12 lazy
    101 trees
      4 drift_detection
      5 evaluation

```
```
# src 
    112 data
     48 meta
      9 utils
      3 metrics
     82 lazy
    187 trees
     37 visualization
     17 drift_detection
      1 options
      6 core
     92 evaluation
     46 demos
      9 bayes
```

I wouldn't mix two styles at the same time (99 and 120 characters). I still can do it in one PR.

**Comment by jacobmontiel** [comment]
Just out of curiosity, how many lines would be affected by setting a limit of 79 characters?

Here is an interesting analysis for the discussion: https://jakevdp.github.io/blog/2017/11/09/exploring-line-lengths-in-python-packages/

I agree that two values are not practical and looking into the numbers above it seems that the 99 limit is manageable. Comments @smastelini @jmread @hmgomes @abifet?

**Comment by smastelini** [comment]
It is a very interesting discussion about line lengths indeed. Although I got used to limiting my scripts' lines to 79 characters, I still believe that 99 is a good balance and would result in less refactoring efforts. Besides, as the article's author points, `sklearn` usually uses more characters per line than the other compared packages. An explanation comes from the fact that this package defines classes with long names, such as `RandomForestClassifier`.

In this sense, `skmultiflow` is much closer to `sklearn` than, let say, `numpy`. Hence, sticking to a limit of 99 characters per line sounds like a good compromise to me.

**Comment by garawalid** [comment]
Here is a summary of the number of lines that violates the rule [E501]:(https://www.flake8rules.com/rules/E501.html)

| File        | case of 79 characters           | case of 99 characters  | case of 120 characters  |
| ------------- |:-------------:| :-----:|:-----:|
| src/neural_networks | 6 | - | - |
| src/data | 283 | 112 | 4 |
| src/transform | 11 | - |
| src/meta | 192 | 48 | - |
| src/utils | 41 | 9 | - |
| src/metrics | 27 | 3 |- |
| src/lazy | 197 | 82 | 16 |
| src/trees | 516 | 187 | 2 | 
| src/visualization | 100 | 37 | - |
| src/drift_detection | 44 | 17 | - |
| src/options | 2 | 1 | - |
| src/core | 24 | 6 | - |
| src/evaluation | 253 | 92 |- |
| src/demos | 68 | 46 | 6 | 
| src/bayes | 19 | 9 |- |
| tests/neural_networks | 4 | 3 |- |
| tests/data | 60 | 16 | 2 | 
| tests/transform | 10 | 1 |- |
| tests/meta | 87  | 11 |- |
| tests/utils | 7 | 2|- |
| tests/metrics | 12 | 5|- |
| tests/lazy | 29 | 12|- |
| tests/trees | 132 |101|- |
| tests/drift_detection | 8 | 4 |
| tests/core | 1 | - |- |
| tests/evaluation | 17 | 5 |
| tests/bayes | 2 | - |- |

**Comment by jacobmontiel** [comment]
I agree, **99** seems the sweet spot considering our codebase and the required effort to enforce it.

**Comment by garawalid** [comment]
99 sounds good for me too. I'll continue the PR. 
Thanks @jacobmontiel, @smastelini

**Comment by garawalid** [comment]
I noticed that linting the project in one shot is quite difficult since we have new PRs and my pace is quite slow. So, I decided to break the linting into different PR. 
The pros of this approach are: 
- No need to rebase every new PR and lint again, 
- Low risk of regression after rebasing.
- Small changed lines per PR and easy to review.
- We can perform the lint in parallel (more contributors, parallel PRs).
- Once a sub-package complies with PEP8, all new PRs that change it should respect that because the check is included in the Azure pipeline.

**Note:** I added a temporary proxy script in `/dev/linter.sh`, it will be deleted once this issue is resolved. 

 
Here is the list of sub-packages to be linted:
- [ ] src/skmultiflow/anomaly_detection (#231)
- [ ] src/skmultiflow/bayes (#231)
- [ ] src/skmultiflow/core (#231)
- [ ] src/skmultiflow/data (#231)
- [ ] src/skmultiflow/demos
- [ ] src/skmultiflow/drift_detection
- [ ] src/skmultiflow/evaluation
- [ ] src/skmultiflow/lazy
- [ ] src/skmultiflow/meta
- [ ] src/skmultiflow/metrics  
- [ ] src/skmultiflow/neural_networks  
- [ ] src/skmultiflow/options 
- [ ] src/skmultiflow/prototype 
- [ ] src/skmultiflow/rule  
- [ ] src/skmultiflow/transform 
- [ ] src/skmultiflow/trees 
- [ ] src/skmultiflow/utils
- [ ] src/skmultiflow/visualization

- [ ] tests/anomaly_detection
- [ ] tests/bayes
- [ ] tests/core
- [ ] tests/data
- [ ] tests/drift_detection
- [ ] tests/evaluation
- [ ] tests/__init__.py
- [ ] tests/lazy
- [ ] tests/meta
- [ ] tests/metrics
- [ ] tests/neural_networks
- [ ] tests/prototype
- [ ] tests/rules
- [ ] tests/transform
- [ ] tests/trees
- [ ] tests/utils

**Comment by smastelini** [comment]
Hi @garawalid, since the work in #231 is almost done and the process should be repeated for the other sub-modules, would you mind describing the steps you followed during this first PR?

Thus, we can get some guidelines for future reference.

**Comment by garawalid** [comment]
@smastelini 
Sorry for the late response. 
1. I use Pycharm to lint the code (Code -> Reformat code).
2. I use [autopep8](https://pypi.org/project/autopep8/) to fix [W293](https://www.flake8rules.com/rules/W293.html).
```bash
$ python3 -m autopep8 --select=W293 --in-place --aggressive src/skmultiflow/options/base_option.py
```
3. I address the rest manually until flake8 returns nothing.
```bash
$ python3 -m flake8 src/skmultiflow/options/
```

**Comment by smastelini** [comment]
Thank you @garawalid, these references are going to be much relevant in the future. We are closing #231, but this issue will be kept open to track the modules that still need the PEP8 style formatting, and also to organize the remaining work.
